File: backend/app/ws/protocol.py
"""WebSocket message protocol between the Angular frontend and the backend.

See ``docs/konzept-mcp-apps.md`` section 5.2 for the full message catalogue
and the rationale (WebSocket transport, ``callId`` correlation).

``UIActionResult`` (the ``tool``/``prompt``/``link``/``intent``/``notify``
payloads sent by the rendered UI resource) is reused as-is from
``mcp_ui_server`` - it is exactly the contract the example server's HTML
speaks (see ``mcp-server-example``).
"""

from __future__ import annotations

from typing import Annotated, Literal, Union

from mcp_ui_server.types import UIActionResult
from pydantic import BaseModel, Field, TypeAdapter

from app.mcp_client.ui import SandboxPolicy

# ---- Client -> Server -------------------------------------------------------


class UserMessageIn(BaseModel):
    type: Literal["user_message"] = "user_message"
    text: str


class UIActionIn(BaseModel):
    type: Literal["ui_action"] = "ui_action"
    callId: str
    action: UIActionResult


ClientMessage = Annotated[Union[UserMessageIn, UIActionIn], Field(discriminator="type")]
client_message_adapter: TypeAdapter[ClientMessage] = TypeAdapter(ClientMessage)


# ---- Server -> Client --------------------------------------------------------


class AssistantMessageOut(BaseModel):
    type: Literal["assistant_message"] = "assistant_message"
    text: str


class ToolCallOut(BaseModel):
    """Transparency: the agent is calling a tool."""

    type: Literal["tool_call"] = "tool_call"
    callId: str
    server: str
    tool: str
    args: dict


class ToolResultOut(BaseModel):
    """Transparency: result of a tool call."""

    type: Literal["tool_result"] = "tool_result"
    callId: str
    server: str
    tool: str
    isError: bool
    summary: str


class ResourcePayload(BaseModel):
    uri: str
    mimeType: str
    text: str | None = None
    blob: str | None = None


class UIResourceOut(BaseModel):
    """A new UI resource to render in a sandboxed iframe."""

    type: Literal["ui_resource"] = "ui_resource"
    callId: str
    server: str
    resource: ResourcePayload
    sandbox: SandboxPolicy


class UIUpdateOut(BaseModel):
    """Push data into an already-rendered iframe (identified by callId)."""

    type: Literal["ui_update"] = "ui_update"
    callId: str
    data: dict


class StatusOut(BaseModel):
    type: Literal["status"] = "status"
    state: Literal["thinking", "idle"]


class NoticeOut(BaseModel):
    """Out-of-band UI actions (``link``/``notify``/``intent``) surfaced as a
    small banner in the chat."""

    type: Literal["notice"] = "notice"
    level: Literal["info", "link"]
    message: str
    url: str | None = None


class ErrorOut(BaseModel):
    type: Literal["error"] = "error"
    message: str


ServerMessage = Union[
    AssistantMessageOut,
    ToolCallOut,
    ToolResultOut,
    UIResourceOut,
    UIUpdateOut,
    StatusOut,
    NoticeOut,
    ErrorOut,
]


def dump_server_message(message: ServerMessage) -> dict:
    return message.model_dump(by_alias=True, exclude_none=True)

File: frontend/src/app/models/protocol.ts
/**
 * WebSocket message protocol shared with the backend.
 *
 * Mirrors `backend/app/ws/protocol.py` (server messages) and
 * `mcp_ui_server.types.UIActionResult` (client/iframe actions). See
 * `docs/konzept-mcp-apps.md` section 5.2 for the authoritative description.
 */

// ---- UIActionResult (emitted by a ui_resource iframe) ----------------------

export interface UIActionResultToolCall {
  type: 'tool';
  messageId?: string | null;
  payload: { toolName: string; params: Record<string, unknown> };
}

export interface UIActionResultPrompt {
  type: 'prompt';
  messageId?: string | null;
  payload: { prompt: string };
}

export interface UIActionResultLink {
  type: 'link';
  messageId?: string | null;
  payload: { url: string };
}

export interface UIActionResultIntent {
  type: 'intent';
  messageId?: string | null;
  payload: { intent: string; params: Record<string, unknown> };
}

export interface UIActionResultNotification {
  type: 'notify';
  messageId?: string | null;
  payload: { message: string };
}

export type UIActionResult =
  | UIActionResultToolCall
  | UIActionResultPrompt
  | UIActionResultLink
  | UIActionResultIntent
  | UIActionResultNotification;

/**
 * Strict validator for messages received from a sandboxed ui_resource
 * iframe (see docs/konzept-mcp-apps.md 7.3, "Schema-Validierung"). Returns
 * `null` for anything that does not match the UIActionResult union exactly
 * so the caller can drop/log it instead of forwarding malformed data to the
 * backend.
 */
export function parseUIActionResult(data: unknown): UIActionResult | null {
  if (typeof data !== 'object' || data === null) {
    return null;
  }
  const { type, payload } = data as Record<string, unknown>;
  if (typeof payload !== 'object' || payload === null) {
    return null;
  }
  const p = payload as Record<string, unknown>;

  switch (type) {
    case 'tool':
      if (typeof p['toolName'] === 'string' && typeof p['params'] === 'object' && p['params'] !== null) {
        return data as UIActionResultToolCall;
      }
      return null;
    case 'prompt':
      if (typeof p['prompt'] === 'string') {
        return data as UIActionResultPrompt;
      }
      return null;
    case 'link':
      if (typeof p['url'] === 'string') {
        return data as UIActionResultLink;
      }
      return null;
    case 'intent':
      if (typeof p['intent'] === 'string' && typeof p['params'] === 'object' && p['params'] !== null) {
        return data as UIActionResultIntent;
      }
      return null;
    case 'notify':
      if (typeof p['message'] === 'string') {
        return data as UIActionResultNotification;
      }
      return null;
    default:
      return null;
  }
}

// ---- Client -> Server --------------------------------------------------------

export interface UserMessageIn {
  type: 'user_message';
  text: string;
}

export interface UIActionIn {
  type: 'ui_action';
  callId: string;
  action: UIActionResult;
}

export type ClientMessage = UserMessageIn | UIActionIn;

// ---- Server -> Client --------------------------------------------------------

export interface AssistantMessageOut {
  type: 'assistant_message';
  text: string;
}

export interface ToolCallOut {
  type: 'tool_call';
  callId: string;
  server: string;
  tool: string;
  args: Record<string, unknown>;
}

export interface ToolResultOut {
  type: 'tool_result';
  callId: string;
  server: string;
  tool: string;
  isError: boolean;
  summary: string;
}

export interface ResourcePayload {
  uri: string;
  mimeType: string;
  text?: string | null;
  blob?: string | null;
}

export interface SandboxPolicy {
  sandboxAttributes: string;
  csp: string;
  trustLevel: 'trusted' | 'untrusted';
}

export interface UIResourceOut {
  type: 'ui_resource';
  callId: string;
  server: string;
  resource: ResourcePayload;
  sandbox: SandboxPolicy;
}

export interface UIUpdateOut {
  type: 'ui_update';
  callId: string;
  data: Record<string, unknown>;
}

export interface StatusOut {
  type: 'status';
  state: 'thinking' | 'idle';
}

export interface NoticeOut {
  type: 'notice';
  level: 'info' | 'link';
  message: string;
  url?: string | null;
}

export interface ErrorOut {
  type: 'error';
  message: string;
}

export type ServerMessage =
  | AssistantMessageOut
  | ToolCallOut
  | ToolResultOut
  | UIResourceOut
  | UIUpdateOut
  | StatusOut
  | NoticeOut
  | ErrorOut;

// ---- Host <-> iframe (postMessage, internal to McpUiFrame) -------------------

/** Sent by the iframe bootstrap script when the document size changes. */
export interface UISizeChangedMessage {
  type: 'ui-size-changed';
  payload: { height: number };
}

/**
 * Sent by the host into the iframe to relay a backend `ui_update`. The
 * bootstrap script re-dispatches this as a `mcp-ui-update` CustomEvent on
 * `document` (see `mcp-server-example`'s templates.py).
 */
export interface UIUpdateMessage {
  type: 'ui-update';
  payload: Record<string, unknown>;
}
